xbmc/addons/binary-addons/BinaryAddonManager.cpp

```diff
@@ -172,105 +413,68 @@ bool CBinaryAddonManager::AddAddonBaseEntry(BINARY_ADDON_LIST_ENTRY& entry)
 }
 
-void CBinaryAddonManager::EnableEvent(const std::string& addonId)
+void CBinaryAddonManager::OnDisabledEvent(const std::string& addonId)
 {
-  CSingleLock lock(m_critSection);
-
-  BinaryAddonBasePtr base;
-  auto addon = m_installedAddons.find(addonId);
-  if (addon != m_installedAddons.end())
-    base = addon->second;
-  else
-    return;
-
-  CLog::Log(LOGDEBUG, "CBinaryAddonManager::%s: Enable addon '%s' on binary addon manager", __FUNCTION__, base->ID().c_str());
-  m_enabledAddons[base->ID()] = base;
-
-  /**
-   * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
-   *
-   * Currently only Screensaver and Visualization use the new way and not need informed.
-   */
+  m_control.SendOutMessage(CBinaryAddonDataProtocol::DISABLE_EVENT, const_cast<char*>(addonId.c_str()), addonId.size()+1);
 }
 
-void CBinaryAddonManager::DisableEvent(const std::string& addonId)
+void CBinaryAddonManager::OnPreInstallEvent(const std::string& addonId)
 {
-  CSingleLock lock(m_critSection);
-
-  BinaryAddonBasePtr base;
-  auto addon = m_installedAddons.find(addonId);
-  if (addon != m_installedAddons.end())
-    base = addon->second;
+  Actor::Message *reply;
+  if (m_control.SendOutMessageSync(CBinaryAddonDataProtocol::PRE_INSTALL_EVENT, &reply, 5000, const_cast<char*>(addonId.c_str()), addonId.size()+1))
+  {
+    bool success = reply->signal == CBinaryAddonDataProtocol::ACC;
+    reply->Release();
+    if (!success)
+    {
+      CLog::Log(LOGERROR, "CBinaryAddonManager::%s - returned error", __FUNCTION__);
+    }
+  }
   else
-    return;
-
-  CLog::Log(LOGDEBUG, "CBinaryAddonManager::%s: Disable addon '%s' on binary addon manager", __FUNCTION__, base->ID().c_str());
-  m_enabledAddons.erase(base->ID());
-
-  /**
-   * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
-   *
-   * Currently only Screensaver and Visualization use the new way and not need informed.
-   */
+  {
+    CLog::Log(LOGERROR, "CBinaryAddonManager::%s - timed out", __FUNCTION__);
+  }
 }
 
-void CBinaryAddonManager::InstalledChangeEvent()
+void CBinaryAddonManager::OnPostInstallEvent(const std::string& addonId, bool update, bool modal)
 {
-  BINARY_ADDON_LIST binaryAddonList;
-  CAddonMgr::GetInstance().GetInstalledBinaryAddons(binaryAddonList);
-
-  CSingleLock lock(m_critSection);
+  MsgAddonPostInstall* msg = new MsgAddonPostInstall;
+  msg->addonID = addonId;
+  msg->update = update;
+  msg->modal = modal;
+  m_control.SendOutMessage(CBinaryAddonDataProtocol::POST_INSTALL_EVENT, &msg, sizeof(MsgAddonPostInstall*));
+}
 
-  BinaryAddonMgrBaseList deletedAddons = m_installedAddons;
-  for (auto addon : binaryAddonList)
+void CBinaryAddonManager::OnPreUnInstallEvent(const std::string& addonId)
+{
+  Actor::Message *reply;
+  if (m_control.SendOutMessageSync(CBinaryAddonDataProtocol::PRE_UNINSTALL_EVENT, &reply, 5000, const_cast<char*>(addonId.c_str()), addonId.size()+1))
```

Comment:
With 5000 the call is given 5 seconds of time, if it is finished faster it returns before, otherwise the wait is finished.